databuilder/query_engines/sqlite.py

```diff
@@ -151,9 +282,101 @@ def engine(self):
         return engine
 
 
-def get_tables(obj):
-    if isinstance(obj, sqlalchemy.sql.ClauseElement):
-        return get_referenced_tables(obj)
+def prepare_query(query):
+    query = define_placeholder_tables(query)
+    query = apply_patient_joins(query)
+    return query
+
+
+def define_placeholder_tables(query):
+    """
+    Where a backend doesn't provide us with a single "patient universe" table guaranteed
+    to contain all patients, we instead generate a "placeholder" table to use instead.
+
+    This function identifies such placeholder tables, generates an appropriate Common
+    Table Expression which defines what the contents of this table should be, and
+    attaches it to the query.
+    """
+    id_column = query.selected_columns[0]
+    placeholder_table = id_column.table
```

Comment:
It's not obvious to me why there is always only one placeholder table in this query that needs replacing.

Reply:
These are created by calling `get_select_query_for_patient_domain()` which we can necessarily only ever do once per SELECT query. The _results_ of that query may then get reified and used in another subsequent query, but at that point the placeholder table has already been dealt with.

Reply:
Oooh. That suggests that there's an implicit invariant here which hadn't occurred to me -- for every non-trivial [1] branch in the at least one [2] of the children is reified.

[1] that is -- those which have non-`Value` children
[2] maybe both?